downloader/artifact_downloader.go

```diff
@@ -17,85 +21,144 @@ const (
 	TargetDir     string
 }
 
-func (ad *ConcurrentArtifactDownloader) DownloadAndSaveArtifacts() ([]ArtifactDownloadResult, error) {
-	if _, err := os.Stat(ad.TargetDir); os.IsNotExist(err) {
-		if err := os.Mkdir(ad.TargetDir, filePermission); err != nil {
+type ConcurrentArtifactDownloader struct {
+	Timeout        time.Duration
+	Logger         log.Logger
+	CommandFactory command.Factory
+}
+
+func NewConcurrentArtifactDownloader(timeout time.Duration, logger log.Logger, commandFactory command.Factory) *ConcurrentArtifactDownloader {
+	return &ConcurrentArtifactDownloader{
+		Timeout:        timeout,
+		Logger:         logger,
+		CommandFactory: commandFactory,
+	}
+}
+
+func (ad *ConcurrentArtifactDownloader) DownloadAndSaveArtifacts(artifacts []api.ArtifactResponseItemModel, targetDir string) ([]ArtifactDownloadResult, error) {
+	if _, err := os.Stat(targetDir); os.IsNotExist(err) {
+		if err := os.Mkdir(targetDir, filePermission); err != nil {
 			return nil, err
 		}
 	}
 
-	return ad.downloadParallel(ad.TargetDir)
+	return ad.downloadParallel(artifacts, targetDir)
 }
 
-func (ad *ConcurrentArtifactDownloader) downloadParallel(targetDir string) ([]ArtifactDownloadResult, error) {
+func (ad *ConcurrentArtifactDownloader) downloadParallel(artifacts []api.ArtifactResponseItemModel, targetDir string) ([]ArtifactDownloadResult, error) {
 	var downloadResults []ArtifactDownloadResult
 
-	jobs := make(chan downloadJob, len(ad.Artifacts))
-	results := make(chan ArtifactDownloadResult, len(ad.Artifacts))
+	jobs := make(chan downloadJob, len(artifacts))
+	results := make(chan ArtifactDownloadResult, len(artifacts))
 
 	for i := 0; i < maxConcurrentDownloadThreads; i++ {
 		go ad.download(jobs, results)
 	}
 
-	for _, artifact := range ad.Artifacts {
+	for _, artifact := range artifacts {
 		jobs <- downloadJob{
 			ResponseModel: artifact,
 			TargetDir:     targetDir,
 		}
 	}
 	close(jobs)
 
-	for i := 0; i < len(ad.Artifacts); i++ {
+	for i := 0; i < len(artifacts); i++ {
 		res := <-results
 		downloadResults = append(downloadResults, res)
 	}
 
 	return downloadResults, nil
 }
 
-func (ad *ConcurrentArtifactDownloader) download(jobs <-chan downloadJob, results chan<- ArtifactDownloadResult) {
-	for j := range jobs {
-		fileFullPath := filepath.Join(j.TargetDir, j.ResponseModel.Title)
+func (ad *ConcurrentArtifactDownloader) downloadFile(targetDir, fileName, downloadURL string) (string, error) {
+	fileFullPath := filepath.Join(targetDir, fileName)
 
-		ctx, cancel := context.WithTimeout(context.Background(), ad.Timeout)
+	ctx, cancel := context.WithTimeout(context.Background(), ad.Timeout)
+
+	downloader := filedownloader.NewWithContext(ctx, retry.NewHTTPClient().StandardClient())
+	err := downloader.Get(fileFullPath, downloadURL)
+	cancel()
+	if err != nil {
+		return "", err
+	}
+	return fileFullPath, nil
+}
+
+func (ad *ConcurrentArtifactDownloader) downloadDirectory(targetDir, fileName, downloadURL string) (string, error) {
```

Comment:
This is intentional: I wanted to achieve a similar [archive extraction that we have in the cache-pull step](https://github.com/bitrise-steplib/steps-cache-pull/blob/master/main.go#L153): [extracting the response body as an input stream](https://github.com/bitrise-steplib/bitrise-step-artifact-pull/pull/70/files#diff-eece178ea711c3c065c9200546f0d598f79ce8dd2ccfef88322fcb3fdac0ab27R117) for the tar command so that we don't need to download the file first.

Reply:
What are the benefits of doing it like that? I mean I can see that we do not need to create a temp dir and clean up after it is done but if I understand correctly we keep the file in memory because we are working with the response body. I think on paper this could cause issues probably only in edge cases when someone wants to cache a multi gigabyte large file.

Reply:
Based on my best knowledge here we are streaming the response body to the tar command, rather than keeping the whole content in memory. Response.Body is an io.reader and not the content.